Bug 1330897 - Update the schema for the main ping

Dexterp37 opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla-services/mozilla-pipeline-schemas/29)
<!-- Reviewable:end -->

georgf commented on telemetry/main.schema.json:
```diff
@@ -868,6 +879,49 @@
+      "type": "array"
+    },
+    "process": {
+      "type": ["object", "array"],
```
When can this ever be an array?

georgf commented on telemetry/main.schema.json:
```diff
@@ -868,6 +879,49 @@
       }
     },
+    "event": {
+      "type": "array"
```
I'll have to file a follow-up bug to define this more strictly.

Dexterp37 replied:
Filed bug 1336104 :-)

georgf commented on telemetry/main.schema.json:
```diff
@@ -868,6 +879,49 @@
+        "events": {
+          "type": "array",
+          "additionalProperties": {
+            "$ref": "#/definitions/event"
```
Are you sure this makes it match the individual entries?
I'd expect this to properties to the "events" definition here.

georgf commented:
Can you also update crash.schema.json with the environment updates here?

Dexterp37 commented:
r? @mreid-moz 

But also two questions :)

1. I've tested the main-ping schema on a Spark cluster, and it validated all the pings. Do I need to test that using Hindsight too?
2. I noticed that using the *validate_main.py* and feeding in a ping with broken *event* data doesn't seem to be caught by this schema. For example,

`"events": { "lol": "rotfl"}`

Doesn't make the validator go crazy. I expected it to. Any idea why it isn't?

mreid-moz commented:
I'm gonna defer to @trink on question #1 (my initial thought is "yes please" for the record), but re: question 2, I believe it's because you're using "additionalProperties" instead of a direct reference to the `processData` struct.

For comparison, take a look at [this small tweak](https://github.com/mreid-moz/mozilla-pipeline-schemas/blob/event_schema_comparison/telemetry/main.schema.json). It properly complains if you give it the wrong structure for `events`.

trink commented:
Well if it doesn't work in HS you will never see your data (but you will get alerts) and since there are differences between the validators yeah that is a step we want in our end to end testing.

Dexterp37 commented:
Thanks to @trink I was able to test the schema for some minutes on a test HS instance.

```
[XXX dexter]$ cat dashboard/analysis.errors.error.tsv
49      main schema validation error: SchemaURI: #/properties/application/properties/buildId Keyword: pattern DocumentURI: #/application/buildId
1       invalid submission: failed to parse offset:0 Invalid value.
8       saved-session schema validation error: SchemaURI: #/properties/clientId Keyword: type DocumentURI: #/clientId
1       crash schema validation error: SchemaURI: #/definitions/addon/properties/version Keyword: type DocumentURI: #/environment/addons/activeAddons/jetpack-extension%40dashlane.com/version
9       main schema validation error: SchemaURI: #/properties/environment/properties/settings/properties/update/properties/enabled Keyword: type DocumentURI: #/environment/settings/update/enabled
2       invalid submission: failed to parse offset:0 The document is empty.
3       main schema validation error: SchemaURI: #/properties/clientId Keyword: type DocumentURI: #/clientId
40      crash schema validation error: SchemaURI: #/properties/environment Keyword: required DocumentURI: #/environment
85      saved-session schema validation error: SchemaURI: #/properties/environment Keyword: required DocumentURI: #/environment
21      main schema validation error: SchemaURI: #/definitions/addon/properties/version Keyword: type DocumentURI: #/environment/addons/activeAddons/jetpack-extension%40dashlane.com/version
21      main schema validation error: SchemaURI: #/properties/environment Keyword: required DocumentURI: #/environment
2       main schema validation error: SchemaURI: #/definitions/addon/properties/name Keyword: type DocumentURI: #/environment/addons/activeAddons/floatnotes%40felix-kling.de/name
1       saved-session schema validation error: SchemaURI: #/properties/payload/properties/log Keyword: type DocumentURI: #/payload/log
1       main schema validation error: SchemaURI: #/properties/environment/properties/system/properties/os/properties/installYear Keyword: type DocumentURI: #/environment/system/os/installYear
1       main schema validation error: SchemaURI: #/definitions/addon/properties/updateDay Keyword: type DocumentURI: #/environment/addons/activeAddons/websecurity%40ati.org/updateDay
35      main schema validation error: SchemaURI: #/properties/environment/properties/settings/properties/isDefaultBrowser Keyword: type DocumentURI: #/environment/settings/isDefaultBrowser
3       saved-session schema validation error: SchemaURI: #/properties/environment/properties/build/properties/vendor Keyword: type DocumentURI: #/environment/build/vendor
1       saved-session schema validation error: SchemaURI: #/properties/environment/properties/settings/properties/isDefaultBrowser Keyword: type DocumentURI: #/environment/settings/isDefaultBrowser
1       main schema validation error: SchemaURI: #/definitions/addon/properties/name Keyword: type DocumentURI: #/environment/addons/activeAddons/quicksearch%40peter.faj/name
1       crash schema validation error: SchemaURI: #/properties/environment/properties/settings/properties/update/properties/enabled Keyword: type DocumentURI: #/environment/settings/update/enabled
```

@mreid-moz , the errors don't seem to relate to the stuff that I've added. So.. r? :)

mreid-moz commented:
> 21      main schema validation error: SchemaURI: #/definitions/addon/properties/version Keyword: type DocumentURI: #/environment/addons/activeAddons/jetpack-extension%40dashlane.com/version
> 35      main schema validation error: SchemaURI: #/properties/environment/properties/settings/properties/isDefaultBrowser Keyword: type DocumentURI: #/environment/settings/isDefaultBrowser

These ones seem worth investigating further (though they don't appear to be regressions from this change). Can we either look into them here or file a follow-up bug?

Dexterp37 commented:
I'm taking a look at them now!

Dexterp37 commented:
@mreid-moz , r?

I've updated the schema to account for the fact the isDefaultBrowser [can be null](https://gecko.readthedocs.io/en/latest/toolkit/components/telemetry/telemetry/data/environment.html) if we fail to obtain the shell service.

I've also removed the *addons* and *profile* sections from *required*, as they are fetched asynchronously and can be missing.

As for the null *version* field for the Dashlane addon, that shouldn't be happening. I tried to install the addon, but it correctly reports the version on my system. I'm filing a bug for further investigations.

This was tested on HS, Spark and Locally. @trink was patient enough to put together a nice debugging plugin that printed a sample ping along with the failures above.

Dexterp37 commented:
Filed bug [1337747](https://bugzilla.mozilla.org/show_bug.cgi?id=1337747) for the addon version issue.

mreid-moz commented on telemetry/crash.schema.json:
```diff
@@ -467,6 +511,7 @@
         }
       },
       "required": [
+        "addons",
```
Should `addons` and `partner` be removed from the `required` list here as well?

mreid-moz commented on telemetry/main.schema.json:
```diff
@@ -267,7 +267,7 @@
             },
             "isDefaultBrowser": {
-              "type": "boolean"
+              "type": ["boolean", "null"]
```
Should we make this change to the `crash` schema as well? (aside: we really should factor out the common ping structure, this repetition is no good).

Dexterp37 replied:
Done!
